clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -1134,9 +1134,22 @@ private CompletableFuture<Void> assignPartitions(
         // Make assignment effective on the client by updating the subscription state.
         updateSubscription(assignedPartitions, false);
 
+        // Pause partitions to ensure that fetch does not start until the callback completes.
+        assignedPartitions.forEach(tp -> subscriptions.pause(tp.topicPartition()));
```

Comment:
Shouldn't we pause only the `addedPartitions`?

Moreover, I wonder if using `pause` is the correct approach. Pausing the newly added partitions will ensure that the consumer does not fetch. However, for the newly added partitions, the consumer may also try to initialize their offsets. See `initWithCommittedOffsetsIfNeeded`. I think that we must prevent this too. Otherwise, there is a race between it and the assigned callback for initializing the offsets. Do you see what I mean?

Reply:
Hey, I totally get it, pause is only preventing fetching but we also need to prevent initializing the offsets as they could change during the ongoing callback, I missed that. I'm still thinking about this, but I see 2 approaches for now:
1. Introducing a new state for the newly added partitions, with changes in the logic to avoid fetch and initialization based on this state (not on pause). Partitions would be in this state while the callback executes, and then move into INITIALIALIZING as if they had been just added. This one is tricky and needs more thinking, it would change several core parts around the subscription state that it is used by many components, so hard to even start to think about the impact. 
2. Continue using the pause mechanism to prevent fetching (not re-inventing the wheel for that), and change the `initWithCommittedOffsetsIfNeeded` for the new consumer only, simply to make sure that it initializes positions only for partitions that are INITIALIZING and not paused (conceptually this new state we have in the new consumer while the callback runs). That would probably achieve what we want, changing only the new consumer, which I like, since this whole situation exists only in it, due to its multi-thread app/background. Still thinking about potential unwanted impact of skipping initializing positions for paused partitions that are INITIALIZING...

Reply:
I don't have a strong opinion on the approach. 1. seems definitely cleaner but I don't know how hard it would be to do it. What I don't like about 2. is that `pause` is usually trigger by the user so, in theory, the user could unpause it instead of us. For the revocation, we added the `pendingRevocation` boolean to `TopicPartitionState`. An alternative to using `pause` would be to add another boolean.

Reply:
I agree with this comment. There is a `KafkaConsumer.paused` method which would show that we had paused the partitions internally, and that seems incorrect in this case. They are logically not paused, but practically we are not fetching records until the callback has been completed.

Reply:
Ok, I do agree on the drawbacks of the pause as it is exposed to the user. I will explore the alternative 1, having a specific state/boolean. My concern was messing too much with the existing subscription state + fetch + offsets but should be fine if I can end up with something isolated similar to the revocation one, that would only apply for the new consumer + new partitions while callback runs.